adapters/mqtt-vertx/src/main/java/org/eclipse/hono/adapter/mqtt/VertxBasedMqttProtocolAdapter.java

```diff
@@ -288,78 +288,75 @@ private void handleEndpointConnectionWithAuthentication(final MqttEndpoint endpo
+            LOG.debug("successfully authenticated device id [{}]", deviceId);
             endpoint.accept(false);
 
             endpoint.publishHandler(message -> {
                 final ResourceIdentifier resource = ResourceIdentifier.fromString(message.topicName());
 
-                if (!validateCredentialsWithTopicStructure(resource, authObject.getTenantId(), logicalDeviceId)) {
+                // check that the device publishes to its device-id
+                if (!validateCredentialsWithTopicStructure(resource, authObject.getTenantId(), deviceId)) {
+
                     // if MQTT client does not conform to the topic structure, close the connection (MQTT has no way for errors)
                     endpoint.close();
                 } else {
-                    publishMessage(endpoint, authObject.getTenantId(), logicalDeviceId, message, resource);
+                    // the payload is ALWAYS published to the deviceId downstream
+                    publishMessage(endpoint, authObject.getTenantId(), deviceId, message, resource);
                 }
             });
 
             endpoint.closeHandler(v -> {
-                LOG.debug("connection closed with client [{}], authId [{}], deviceId [{}]",
-                        endpoint.clientIdentifier(), authObject.getAuthId(), logicalDeviceId);
+                LOG.debug("connection closed with tenantId [{}], authId [{}], deviceId [{}]",
+                        authObject.getTenantId(), authObject.getAuthId(), deviceId);
                 if (registrationAssertions.remove(endpoint) != null)
-                    LOG.trace("removed registration assertion for client [{}]", endpoint.clientIdentifier());
+                    LOG.trace("removed registration assertion for tenantId [{}], authId [{}], deviceId [{}]",
+                            authObject.getTenantId(), authObject.getAuthId(), deviceId);
             });
 
         } else {
-            LOG.debug("authentication failed for client [{}] rejected [{}]",
-                    endpoint.clientIdentifier(), MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED);
+            LOG.debug("authentication failed for tenantId [{}], authId [{}] rejected [{}]",
+                    authObject.getTenantId(), authObject.getAuthId(), MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED);
             endpoint.reject(MqttConnectReturnCode.CONNECTION_REFUSED_NOT_AUTHORIZED);
         }
     }
 
-    private void publishMessage(final MqttEndpoint endpoint, final String tenantId, final String logicalDeviceId,
+    private void publishMessage(final MqttEndpoint endpoint, final String tenantId, final String deviceId,
             final MqttPublishMessage message, final ResourceIdentifier resource) {
-        LOG.trace("received message [client ID: {}, topic: {}, QoS: {}, payload {}]", endpoint.clientIdentifier(),
+        LOG.trace("received message [tenantId [{}], deviceId [{}], topic: [{}], QoS: [{}], payload [{}]]", tenantId, deviceId,
```

Comment:
no need to nest brackets here, i.e.
`received message [tenant-id: {}, device-id: {}, ... ]` will do fine.

Reply:
would keep the brackets around payload though (could be quite awful to read otherwise?)